src/relations/backend_database.py

```diff
@@ -69,45 +82,166 @@ def __init__(self, charm: CharmBase):
         """
+        logger.info("initialising postgres and pgbouncer relations")
+        if self.postgres is None:
+            event.defer()
+            logging.error("deferring database-created hook - postgres database not ready")
+            return
+
+        plaintext_password = pgb.generate_password()
+        # create authentication user on postgres database, so we can authenticate other users
+        # later on
+        self.postgres.create_user(self.auth_user, plaintext_password, admin=True)
+        self.initialise_auth_function(dbname=self.database.database)
+
+        hashed_password = pgb.get_hashed_password(self.auth_user, plaintext_password)
+        self.charm.push_file(
+            f"{PGB_DIR}/userlist.txt", f'"{self.auth_user}" "{hashed_password}"', perms=0o400
+        )
+        cfg = self.charm.read_pgb_config()
+        # adds user to pgb config
+        cfg.add_user(user=event.username, admin=True)
+        cfg["pgbouncer"][
+            "auth_query"
+        ] = f"SELECT username, password FROM {self.auth_user}.get_auth($1)"
+        cfg["pgbouncer"]["auth_file"] = f"{PGB_DIR}/userlist.txt"
+        self.charm.render_pgb_config(cfg)
+
+        self.charm.update_postgres_endpoints(reload_pgbouncer=True)
+
+    def _on_endpoints_changed(
+        self, _: Union[DatabaseEndpointsChangedEvent, DatabaseReadOnlyEndpointsChangedEvent]
+    ):
+        self.charm.update_postgres_endpoints(reload_pgbouncer=True)
+
+    def _on_relation_departed(self, event: RelationDepartedEvent):
+        """Runs pgbouncer-uninstall.sql and removes auth user.
+
+        pgbouncer-uninstall doesn't actually uninstall anything - it actually removes permissions
+        for the auth user.
+        """
+        if event.departing_unit != self.charm.unit:
+            return
+
+        logger.info("removing auth user")
+
+        uninstall_script = open("src/relations/sql/pgbouncer-uninstall.sql", "r").read()
+
         try:
-            self.charm.add_user(
-                user=event.username,
-                cfg=self.charm.read_pgb_config(),
-                password=event.password,
-                admin=True,
-                reload_pgbouncer=True,
-                render_cfg=True,
+            with self.postgres.connect_to_database(PGB_DB) as conn, conn.cursor() as cursor:
+                cursor.execute(uninstall_script.replace("auth_user", self.auth_user))
+            conn.close()
+        except psycopg2.Error:
+            self.charm.unit.status = BlockedStatus(
+                "failed to remove auth user when disconnecting from postgres application."
             )
+            return
 
-            self.charm.update_postgres_endpoints()
+        self.postgres.delete_user(self.auth_user)
 
-        except FileNotFoundError:
-            event.defer()
+        logger.info("auth user removed")
 
-    def _on_relation_broken(self, event: RelationBrokenEvent):
+    def _on_relation_broken(self, _: RelationBrokenEvent):
```

Comment:
If you are not caring about the event, you should not care about it's type either, i.e.:
```suggestion
    def _on_relation_broken(self, _):
```